### PR TITLE
Refactor instrument  module

### PR DESCRIPTION
**Pull request by MaxNoe**
Working on #971 

* Only guesses telescope names, not full camera geometries or optics
* Refactoring of the members of `TelescopeDescription`, `OpticsDescription`
* Make those classes hashable, so set operations work and they can be used as dict keys

### CODE REVIEW

**Comment by dneise** [comment]
Btw, I was curious if one can find the telescope names somewhere in the eventio file, so that guess could potentially avoided completely. It found them in the "History" part, I think, but only in comments looking like this:

```
ECHO ************************************************************************
ECHO Configuration for single-reflector SST-1M 5 in Prod-4.
ECHO ************************************************************************
ECHO 
ALTITUDE 2000. % m (Typically replaced on the command line).
ATMOSPHERIC_TRANSMISSION atm_trans_2000_1_10_0_0_2000.dat
TRIGGER_TELESCOPES 1           % Unless an array trigger file sets up more specific triggers.
[.. more configuration ..]
```

I guess this is not a good way to read the telescope name from the file, but I wanted to mention it anyway.

**Comment by MaxNoe** [comment]
The tests now failing are due to  testing for magic numbers in the old simtel file.

Any advice on how to replace those values?

**Comment by mackaiver** [comment]
> The tests now failing are due to testing for magic numbers in the old simtel file.
> 
> Any advice on how to replace those values?

If we find no other way to check for correct values then these tests can be removed imo.

**Comment by kosack** [comment]
>I guess this is not a good way to read the telescope name from the file, but I wanted to mention it anyway.

I know that I had mentioned that it would be nice to have the names to Konrad, and I think he said he would add that in a future version. It may be there already somewhere. In any case, we still probably want to map names there to "standard" ones we define, so there is no confusion.

**Comment by kosack** [comment]
> If we find no other way to check for correct values then these tests can be removed imo.

was the example file updated? If so, yes we should remove all tests that assume specific values. It's better to give the tests a hand-made dummy event where we know the correct outcome, rather than using a real event (which can be tested at a higher level if necessary).  Probably that should be a separate PR.

**Comment by kosack** [comment]
Maybe not necessary for this PR, but it would be nice to add a "guess" function for the SubarrayDescription itself, so that at least the Prod ID and the array layout could be shown (not sure if that is info directly in the MC headers or not). Right now all subarrays are called "MonteCarloArray", which is not so useful to identify them.

**Comment by MaxNoe** [comment]
> I know that I had mentioned that it would be nice to have the names to Konrad, and I think he said he would add that in a future version. It may be there already somewhere. In any case, we still probably want to map names there to "standard" ones we define, so there is no confusion.

I just produced small test files using the current developement version of simtel_array (prod4 baselines la palma and paranal) and I found no changes in the containers that describe the telescopes.

**Comment by MaxNoe** [comment]
> was the example file updated? If so, yes we should remove all tests that assume specific values. It's better to give the tests a hand-made dummy event where we know the correct outcome, rather than using a real event (which can be tested at a higher level if necessary). Probably that should be a separate PR.

Yes, I changed the `example_event` to come from the prod3 `gamma_test_large.simtel.gz`

**Comment by MaxNoe** [comment]
I removed the magic number tests, test pass now.
I opened an issue here: #988 to get real tests for those features.

**Comment by MaxNoe** [comment]
Tests are passing, please review @watsonjj @kosack @dneise @ParsonsRD

File: ctapipe/instrument/guess.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+GuessingKey = namedtuple('GuessingKey', ['n_pixels', 'focal_length'])
+GuessingResult = namedtuple(
+    'GuessingResult', ['type', 'name', 'camera_name', 'n_mirrors']
+)
+
+
+TELESCOPE_NAMES = {
+    GuessingKey(2048, 2.28): GuessingResult('SST', 'GCT', 'CHEC', 2),
+    GuessingKey(2368, 2.15): GuessingResult('SST', 'ASTRI', 'ASTRICam', 2),
```
Comment (watsonjj) [inline ctapipe/instrument/guess.py]: Can an entry be added for:
`GuessingKey(2048, 2.15): GuessingResult('SST', 'ASTRI', 'CHEC', 2),`

**Comment by mackaiver** [comment]
So the difference between the two test fles is the MC production? Prod2 vs Prod3? Maybe we should rename that test file to something conveying more information.

**Comment by MaxNoe** [comment]
Yes, but this would be fore ctapipe-extra. I can also produce more sensible test files not, I got a recent simtel running.

File: ctapipe/instrument/guess.py (hunk 1 above)
Comment (MaxNoe) [reply ctapipe/instrument/guess.py]: done

**Comment by MaxNoe** [comment]
Please review again, I added @watsonjj comment.

**Comment by kosack** [comment]
> Yes, but this would be fore ctapipe-extra. I can also produce more sensible test files not, I got a recent simtel running.

It would be nice to have some smaller nicer test files and get rid of the old ones in ctapipe_extra which is becoming too large already: at least something like 10+ gamma rays for north and south configurations, and some muon-only files. For deeper/larger test files, we can just make them automatically download from the server, rather than put them in the package.   In the future, we will move to selectable configurations as in #738, so this is just a temporary fix.

**Pull request merged**